### PR TITLE
Prep changes for next version of SAM

### PR DESCRIPTION
**Pull request by abrooksv**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description
Prep for SAM 0.30.0

* Refactor the template parameters to not rely on mutually exclusive nulls

## Testing
None. There is a bug in SAM preventing testing. Don't merge yet, but feel free to review.

## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by abrooksv** [comment]
@awood45 Changes made to JetBrains

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/SamInitRunner.kt [hunk 1]
```diff
@@ -9,46 +9,59 @@ import com.intellij.openapi.vfs.VfsUtil
...
+                            .withParameters(name)
+                            .withParameters("--runtime")
+                            .withParameters(runtime.toString())
+                            .withParameters("--dependency-manager")
+                            .withParameters(templateParameters.dependencyManager)
 
-                dependencyManager?.let {
-                    this.withParameters("--dependency-manager")
-                        .withParameters(it)
+                        if (isSamGte30) {
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/SamInitRunner.kt]: Like the previous SAM change like this, we should add a `FIX_WHEN_SAM_MIN_IS_0_30` or similar

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProjectWizard.kt [hunk 2]
```diff
@@ -61,6 +61,11 @@ data class SdkBasedSdkSettings(
     val sdk: Sdk?
 ) : SdkSettings
 
+sealed class TemplateParameters {
+    data class LocationBasedTemplate(val location: String) : TemplateParameters()
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProjectWizard.kt]: shouldn't this own which parameters are included? that means we can more easily add new types.

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProjectWizard.kt]: I couldn't decide if I wanted all CLI generation in 1 spot using a sealed class, vs abstract method on TemplateParameter that takes a GeneralCommandLine which makes it harder to find what is the concrete CLI generated.... So went with code readability over flexibility. I dont see SAM adding a 3rd way anytime soon.

**Pull request merged**